codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java

```diff
@@ -895,111 +910,53 @@ public String toString() {
         return builder.append(']').toString();
     }
 
-    protected ValueConverter<T> valueConverter() {
-        return valueConverter;
-    }
-
-    @SuppressWarnings("unchecked")
-    private void updateSizeAfterSet(Object oldValue, int newSize) {
-        if (isList(oldValue)) {
-            size -= ((List<T>) oldValue).size();
-        } else if (oldValue != null) {
-            size--;
-        }
-        size += newSize;
-    }
-
-    private void validateName(T name) {
-        checkNotNull(name, "name");
-        nameValidator.validate(name);
+    protected HeaderEntry newHeaderEntry(int h, T name, T value, HeaderEntry next) {
+        return new HeaderEntry(h, name, value, next);
     }
 
-    private static boolean isList(Object value) {
-        return value != null && value.getClass().equals(ValuesList.class);
-    }
-
-    private List<T> newList() {
-        return newList(4);
+    protected ValueConverter<T> valueConverter() {
+        return valueConverter;
     }
 
-    private List<T> newList(int initialSize) {
-        return new ValuesList<T>(initialSize);
+    private void add0(int h, int i, T name, T value) {
+        // Update the hash table.
+        entries[i] = newHeaderEntry(h, name, value, entries[i]);
+        ++size;
     }
 
-    private final class NameValueIterator implements Iterator<Entry<T, T>> {
-        private final Iterator<Entry<T, Object>> iter;
-        private T name;
-        private List<T> values;
-        private int valuesIdx;
-
-        NameValueIterator(Iterator<Entry<T, Object>> iter) {
-            this.iter = iter;
-        }
-
-        @Override
-        public boolean hasNext() {
-            return iter.hasNext() || values != null;
+    private void remove0(int h, int i, T name) {
```

Comment:
The first loop can update the `next` links and also the `entries` array. The second loop only updates the `next` links because we know there is at least one entry that won't be removed in the `entries[i]` bucket.

Did you have something simpler in mind?